src/Hazelcast.Net/Clustering/ClusterConnections.cs

```diff
@@ -889,7 +896,23 @@ private async Task<MemberConnection> ConnectAsync(NetworkAddress address, Cancel
             // about it if adding support for parallel connections!
 
             // connection is opened
-            await RaiseConnectionOpened(connection, isFirstEver, isFirst, isNewCluster).CfAwait();
+            try
+            {
+                await RaiseConnectionOpened(connection, isFirstEver, isFirst, isNewCluster).CfAwait();
```

Comment:
Confused a bit here. We set completion null in either case here and just below in the locked scope after `RaiseConnectionOpened`. Isn't this should set exception instead? I see that we should await here before throw but I couldn't predict the result of setting null in both cases.

Reply:
if `RaiseConnectionOpened` fails, we're going to trash the connection and throw - but before that we want to get rid of the completion. OTOH if it does not fail, we also need to remove the completion. So it's either (a) `RaiseConnectionOpened` succeeds, we enter the mutex block and remove the completion, or (b) it fails, we remove the completion, and throw.

We cannot remove the completion in a `finally` block because we need to do it *before* we dispose the connection.

Now... what strucks me is... why use a mutex in one case and not the other? Thinking about it...

Reply:
I think that's okay since we expect that exception is not likely and operation is just removing from the dictionary.